### PR TITLE
Create, Update, and Delete operations on the Calendar and its events. Along with various CSS changes.

### LINKED ISSUE #37: [ISSUES] Need to have the ability to create, update, and delete events for calendar.
**Description of the issue**
Describe the issue and share important details such as the desired or expected solution and error messages. Specify where the issue originates and whether it affects the frontend, backend, or both. 
This issue will require both the backend and frontend to be updated. There will need to be buttons in the frontend calendar where when clicked, the user will have the ability to create, update, and delete events. 
The user will only have the ability to create, read, update, and delete events they have permissions to, so there needs to be an association between the current user logged into the browser and the aforementioned operations. 
There also needs to be modals or forms with input boxes for the user to edit the details of events when they create and update events.
There needs to be checks and error messages when users create or update events with invalid details (such as a missing title or start time).

**Files that need to be updated**
List all files or directories that need to be changed/updated to resolve the issue. Include any relevant details to ensure the changes are necessary.
The following backend files will need to be updated:
backend\calendar_app\serializers.py
backend\calendar_app\models.py
backend\calendar_app\views.py
backend\calendar_app\urls.py
Any files inside the following folder will need to be updated, as well as potentially new files:
frontend\src\components\Calendar\ 

**Link**
(Pull request will be linked later)

### PR DESCRIPTION
**Pull request by nnguyen144**
### Category
- [X] **Feature**
- [ ] **Bug Fix**
- [X] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [ ] **Frontend**
- [ ] **Backend**
- [X] **Both**

### Description
This PR adds functionality in the backend and frontend to create, delete, and update events for display on the calendar.

### Related Issue(s)
Issue #37 

### New Additions
4 new files were added: UpdateForm.js, CreateForm.js, and their respective CSS files.
These additions are necessary as there needs to be a modal that displays when a user clicks a button to create or update an event. 

### Modifications
Files modified: models.py, urls.py, views.py, serializers.py, tests.py, Calendar.js, Calendar.module.css, EventModal.js, EventModal.module.css, page.js, and two migration files.
The backend files were updated to handle the CRUD operations, and the frontend files were updated to allow users to perform CRUD operations on the browser.

### Visuals
The following is a gif showcasing how one can create a personal event. 
opera_Mw8QCD0hxl

The following is a gif showcasing how one can create a hub event.
opera_ZjSieQ3zrB

The following is a gif showcasing how one can update an event.
opera_dZcFjskRAb

The following is a gif showcasing how one can delete an event.
opera_EsnYrWnMij

The following is a gif showcasing how one can use the month dropdown.
opera_HORbVCLsnt

### Architecture of Your Changes
This PR adds CRUD operations for the events on the calendar, as well as other additions. This includes the following:

-  Create button on the top of the calendar where, when clicked, a modal will open that allows users to input a title, location, description, start_time, end_time, hub, isPersonal boolean, and color for a new event. The isPersonal is inputted via a checkbox, the start_time and end_time are inputted via an interactive mini calendar, the hub is inputted via a dropdown list (and only displays hubs which the current user is owner of or a moderator of), and the color is inputted via a color picker.
- Dropdown menu when the month is clicked on the top right of the calendar to display a dropdown list of months in order to quickly switch month.
- Update button on the bottom left of an event modal where, when clicked, a modal will open that allows users to input a title, location, description, start_time, end_time, and color for the current event.
- Delete button on the bottom right of an event modal where, when clicked, the event modal will close and the event is deleted from the calendar.

Additions were made in the backend to support this, including API calls to create, update, and delete events. 
Tests were included in order to test the robustness of the backend.

### Usage and Integration
The calendar is a standalone page, so it does not interact with the other parts of this project other than there needing to be a button (preferably on a sidebar or header) that directs the user to the calendar.
Since events can be tied to hubs, events for a hub can be shown on a hub page.

### Testing
- **Testing Approach**: I've applied integration testing.
- **Edge Cases**: Edge cases I've addressed include bad data inputs when creating and deleting an event. These include, but potentially are not limited to, missing title, missing start time, start time being later than end time, either both or neither of the hub and personal event options are chosen, creating an event for a hub without valid permissions (owner/moderator), and invalid color hex code.
- **Test Commands**: I used "python manage.py calendar_app.test" in order to test my code.

### CODE REVIEW

File: frontend/src/components/Calendar/Calendar.js [hunk 1]
```diff
@@ -24,19 +39,92 @@ const Calendar = () => {
...
+      console.log("Error deleting event: ", error.response);
+    } 
+  }
+
+  // Fetches events and hubs from API
   useEffect(() => {
-    console.log("Fetching...");
+
+    // Need to fix this. Works if first time trying to access website, but doesn't work for subsequent uses.
+    const token = localStorage.getItem(ACCESS_TOKEN);
```
Comment (thecatfour) [inline frontend/src/components/Calendar/Calendar.js]: If this is something that you cannot fix currently, I think you should create a new issue to remind us to fix it in the future.

File: frontend/src/components/Calendar/Calendar.js [hunk 2]
```diff
@@ -24,19 +39,92 @@ const Calendar = () => {
...
+    if(!token){
+      console.log("User is currently not logged in. Redirecting to login page.");
+      router.push('/users/login');
+      return;
+    }
+
+    console.log("Fetching the hubs the current user is a owner of or is moderating...");
+    const fetchHubs = async () => {
+      try {
+          const response = await api.get("/api/hubs/modding/", {headers: {Authorization: `Bearer ${token}`}});
```
Comment (thecatfour) [inline frontend/src/components/Calendar/Calendar.js]: I don't think specifying the authorization token is necessary when using our api utility when in a ProtectedRoute. For example, the page that displays a post list currently uses this: 

image

Maybe I'm overlooking something, but I thought the api component added the token by itself.

**Comment by thecatfour** [comment]
Overall, I think this PR can be merged. 

If there are any bugs or issues, then they should be made into issues on the repository and linked here to show that they have been made.

**Review by thecatfour — DISMISSED** [review]

**Comment by courtmcgee2197** [comment]
So after testing everything it should be working however, I do see you are adding tokens to your api calls which is implicitly handled by the api function. When I tested on my end I commented out the token portions and I did not have any issues with calendar and I changed between different users to test it as well changed the expiration of tokens to make sure they were working properly. There should be no calling of tokens outside of the code the api.js and protectedRoute.js as they already handle this aspect. 

After removing the code dealing with tokens I do believe this can be merged as it worked on my end without any logic regarding tokens.

File: frontend/src/components/Calendar/Calendar.js (hunk 2 above)
Comment (nnguyen144) [reply frontend/src/components/Calendar/Calendar.js]: You are right about this. I removed the authorization token and the output is the same. I will remove it and update it.

**Comment by nnguyen144** [comment]
\> After removing the code dealing with tokens I do believe this can be merged as it worked on my end without any logic regarding tokens.

You are right. I will remove them.

File: frontend/src/components/Calendar/Calendar.js (hunk 1 above)
Comment (nnguyen144) [reply frontend/src/components/Calendar/Calendar.js]: I will remove it for now.

**Review by courtmcgee2197 — APPROVED** [review]
After removing the token logic this is safe for merge

**Review by thecatfour — APPROVED** [review]
Everything seems to be good to merge.

**Pull request merged**